### ISSUE TITLE
Be less dependent on axios

### ISSUE THREAD

**Issue by simplesmiler** [simplesmiler/taxios#4]
# TL;DR
Current problems are:
1. We can not be independent of axios, because we need `AxiosResponse` generic.
2. Axios can not be peer dependency because TypeScript does not work well with peer dependencies.

# Details

To be independent of axios, we have to be able to arbitrarily wrap response types (e.g. `Pet` into `AxiosResponse<Pet>`).

But TypeScript does not support:
1. First class generics, e.g. generic as parameter for another generic.
    ```ts
    type Wrap<T> = { data: T };
    const taxios = new Taxios<Api, Wrap>(agent);
    ```
2. `ReturnType` of generic function.
    ```ts
    type Wrap<T> = { data: T };
    type Wrapper = <T>() => Wrap<T>;
    type WrappedNumber = ReturnType<Wrapper<number>>;
    ```

So far I could not find a way around that.

**Comment by simplesmiler** [comment]
Maybe this still possible with classes, because classes can have generic methods and can themselves be generic parameters.

But maybe we don't want per-agent response type at all, but always want native `Response` type branded with T?

**Comment by crutch12** [comment]
> But maybe we don't want per-agent response type at all, but always want native `Response` type branded with T?

Native `Response` is too uncomfortable to work with. I think this is the reason why people use axios.

If you want to reject axios, I think we should create another npm module for that

**Comment by simplesmiler** [comment]
TLDR: I think axios is slowly becoming a "jQuery".

I think people use axios mostly because:
- XMLHttpRequest is awful to work with. Fetch feels much better in comparison (but not perfect), and axios feels even better (but at a cost of not supporting web streaming).
- Axios has interceptors for "global" behaviors, like logging, authentication and retry.

> Native `Response` is too uncomfortable  to work with.

I would like to argue that it's not really. Response and AxiosResponse have mostly the same interface, with only notable exceptions being:
- AxiosResponse has `config: AxiosConfig`, which is mostly used in interceptors.
- Response is returned before all data has arrived, so it has to be `{ json(): Promise<T> }` instead of `{ data: T }`. But I personally don't see much difference between `taxios.get('/users').then(r => r.data)` and `taxios.get('/users').then(r => r.json())`.

**Comment by simplesmiler** [comment]
Judging by https://github.com/axios/axios/issues/4209, there is intent in axios to support web streams.
